File: src/main/java/com/baidu/algorithm/backtracking/_78_Subsets.py
class Solution(object):

    def dfs(self, res, one_res, nums, index):
        res.append(one_res[:])

        for i in range(index, len(nums)):
            one_res.append(nums[i])
            self.dfs(res, one_res, nums, i + 1)
            one_res.pop()

    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if nums is None or len(nums) == 0:
            return [[]]

        nums.sort()
        res = []

        self.dfs(res, [], nums, 0)

        return res

File: src/main/java/com/baidu/algorithm/backtracking/_78_Subsets.java
/**
 * Copyright (C) 2016 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.backtracking;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * _78_Subsets
 *
 * @author xuhaoran01
 */
public class _78_Subsets {

    private void dfs(List<List<Integer>> res, List<Integer> oneRes, int[] nums, int index) {

        res.add(new ArrayList<>(oneRes));

        for (int i = index; i < nums.length; i++) {
            if (i > index && nums[i] == nums[i - 1]) {
                continue;
            }

            oneRes.add(nums[i]);
            dfs(res, oneRes, nums, i + 1);
            oneRes.remove(oneRes.size() - 1);
        }
    }

    public List<List<Integer>> subsets(int[] nums) {

        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }

        Arrays.sort(nums);

        dfs(res, new ArrayList<>(), nums, 0);

        return res;
    }
}
